### PR TITLE
Make all float->int/frac conversions round to nearest by default

### LINKED ISSUE #1931: pio_calculate_clkdiv_from_float is not rounding
The `pio_calculate_clkdiv_from_float()` function is truncating the fractional result:

```c
*div_frac = (uint8_t)((div - (float)*div_int) * (1u << 8u));
```

It should be rounding. One way to do this is to add half of 1/256 up front, as the very first line of the function, before the range check:

```c
// Bump div by 1/512 so that we're rounding instead of truncating.
div += 0.001953125f;
```

This error caused an NTSC color burst to be far enough off in frequency to not be recognized by two different monitors.

### PR DESCRIPTION
**Pull request by kilograham**
Under control of specific vars, but defaulted by `PICO_CLKDIV_ROUND_NEAREST` which now defaults to `1`

This fixes #1931 which was PIO specific, but it seemed sensible to be consistent across all APIs

### CODE REVIEW

File: src/rp2_common/hardware_pio/include/hardware/pio.h [hunk 1]
```diff
@@ -488,6 +494,9 @@ static inline void sm_config_set_clkdiv_int_frac(pio_sm_config *c, uint16_t div_
 
 static inline void pio_calculate_clkdiv8_from_float(float div, uint32_t *div_int, uint8_t *div_frac8) {
     valid_params_if(HARDWARE_PIO, div >= 1 && div <= 65536);
+#if PICO_PIO_CLKDIV_ROUND_NEAREST
+    div += 0.5f / 256; // round to the nearest 1/256
```
Comment (lurch) [inline src/rp2_common/hardware_pio/include/hardware/pio.h:498]: I guess the hardcoded `256` here could be `(1 << (PIO_SM0_CLKDIV_FRAC_MSB + 1 - PIO_SM0_CLKDIV_FRAC_LSB))` for consistency with the other changes in this PR?

Comment (kilograham) [reply src/rp2_common/hardware_pio/include/hardware/pio.h:498]: true, but i decided not to

File: src/rp2_common/hardware_adc/include/hardware/adc.h [hunk 2]
```diff
@@ -198,6 +203,9 @@ static inline void adc_run(bool run) {
  */
 static inline void adc_set_clkdiv(float clkdiv) {
     invalid_params_if(HARDWARE_ADC, clkdiv >= 1 << (ADC_DIV_INT_MSB - ADC_DIV_INT_LSB + 1));
+#if PICO_ADC_CLKDIV_ROUND_NEAREST
+    clkdiv += 0.5f / (1 << ADC_DIV_INT_LSB); // round to the nearest fraction
```
[the remark this answers is not in the document]
Comment (kilograham) [reply src/rp2_common/hardware_adc/include/hardware/adc.h:207]: true but i decided not to

[the remark this answers is not in the document]
Comment (lurch) [reply src/rp2_common/hardware_adc/include/hardware/adc.h:207]: Fair enough!

**Comment by lurch** [comment]
All looks sensible, and `tools/extract_configs.py` is happy with all the new `PICO_CONFIG:` additions :+1:

**Review by peterharperuk — DISMISSED** [review]

File: src/rp2_common/hardware_pwm/include/hardware/pwm.h [hunk 3]
```diff
@@ -155,6 +160,9 @@ static inline void pwm_config_set_phase_correct(pwm_config *c, bool phase_correc
  */
 static inline void pwm_config_set_clkdiv(pwm_config *c, float div) {
     valid_params_if(HARDWARE_PWM, div >= 1.f && div < 256.f);
+#if PICO_PWM_CLKDIV_ROUND_NEAREST
+    div += 0.5f / (1 << PWM_CH0_DIV_INT_LSB); // round to the nearest fraction
```
[the remark this answers is not in the document]
Comment (lurch) [reply src/rp2_common/hardware_pwm/include/hardware/pwm.h:164]: That's (almost) exactly what I asked earlier - see my resolved comments above.

[the remark this answers is not in the document]
Comment (peterharperuk) [reply src/rp2_common/hardware_pwm/include/hardware/pwm.h:164]: true

[the remark this answers is not in the document]
Comment (lurch) [reply src/rp2_common/hardware_pwm/include/hardware/pwm.h]: Done :slightly_smiling_face:

File: src/rp2_common/hardware_pio/include/hardware/pio.h [hunk 4]
```diff
@@ -495,7 +495,7 @@ static inline void sm_config_set_clkdiv_int_frac(pio_sm_config *c, uint16_t div_
 static inline void pio_calculate_clkdiv8_from_float(float div, uint32_t *div_int, uint8_t *div_frac8) {
     valid_params_if(HARDWARE_PIO, div >= 1 && div <= 65536);
 #if PICO_PIO_CLKDIV_ROUND_NEAREST
-    div += 0.5f / 256; // round to the nearest 1/256
```
Comment (peterharperuk) [inline src/rp2_common/hardware_pio/include/hardware/pio.h]: I preferred the constant!

File: src/rp2_common/hardware_pwm/include/hardware/pwm.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (peterharperuk) [reply src/rp2_common/hardware_pwm/include/hardware/pwm.h]: wish I hadn't asked!

File: src/rp2_common/hardware_pio/include/hardware/pio.h (hunk 4 above)
Comment (lurch) [reply src/rp2_common/hardware_pio/include/hardware/pio.h]: :man_facepalming:  You're going to _love_ the additional commit that I've got coming soon then... :cold_sweat:

**Comment by lurch** [comment]
Latest commit just pushed. Let me know if it goes "too far" ( :face_with_peeking_eye: ) and I'll revert it and push a simpler version without the `field_width` stuff.

**Comment by kilograham** [comment]
nice; i think I'd prefer REG_FIELD_WIDTH() and as such it probably belongs in `platform_defs.h`. You'll have to duplicate it in both copies for now (maybe we'll extract a common header in the future)

**Comment by lurch** [comment]
> i think I'd prefer REG_FIELD_WIDTH() and as such it probably belongs in `platform_defs.h`

Done! :+1:

**Review by kilograham — COMMENTED** [review]
awesome

**Review by peterharperuk — APPROVED** [review]

**Review by peterharperuk — APPROVED** [review]

**Pull request merged**